index.js

```diff
@@ -835,6 +856,89 @@ class Batch {
   }
 }
 
+class Watcher extends EventEmitter {
+  constructor (bee, range) {
+    super()
+
+    this.bee = bee
+    this.core = bee.core
+
+    this.watching = false
+    this.running = false
+    this.destroyed = false
+
+    this.range = range
+    this.latestDiff = this.bee.version
+    this.stream = null
+
+    const onappend = this._onappend.bind(this)
+    this.onappend = debounceify(onappend)
+
+    // this.on('error', noop)
+  }
+
+  async _onappend () {
+    try {
+      await this._run()
+    } catch (err) {
+      this.emit('error', err)
+      this.destroy()
+    } finally {
+      this.running = false
+    }
+  }
+
+  async _run () {
+    this.running = true
+
+    // + I think it could repeat a last call even after bee is closed, should trigger it with a test first
```

Comment:
Yeah, indeed, and there's another race condition too. Not sure if we need to care about these, but they could lead to some confusion:
- A watcher can be unwatched/destroyed while an onAppend is being processed. So in that case too you would observe an event after closing. I believe the solution there is to make this.running a promise or null, and adding
`if (this.running) await this.running` after setting this.watching to false

Reply:
Ah, just understood your logic with storing this.stream as instance variabele. That also solves the race condition I described, and probably in a cleaner way (it won't be forced to wait until all diffs are yielded)

Reply:
I can reproduce it by manually adding a delay before the diff stream loop (i.e. simulates network)

I couldn't reproduce it on a test but it's obviously possible. We don't want to trigger any `'change'` event or any error if the user already manually did `destroy()` so I think we can safely return right there

Reply:
I think this could be solved if we could cancel the next debounce call if any, but it's much simpler to just check if destroyed and stop the execution as we're doing now

Reply:
I didn't think it was possible to cancel something which has been scheduled on the event loop. Is it?

Reply:
I meant the `debounceify` debounced last call, but overkill